Kernel/Interrupts/InterruptManagement.cpp

```diff
@@ -0,0 +1,229 @@
+#include <AK/FixedArray.h>
+#include <Kernel/ACPI/MultiProcessorParser.h>
+#include <Kernel/Arch/i386/CPU.h>
+#include <Kernel/Interrupts/IOAPIC.h>
+#include <Kernel/Interrupts/InterruptManagement.h>
+#include <Kernel/Interrupts/PIC.h>
+#include <Kernel/VM/MemoryManager.h>
+
+#define PCAT_COMPAT_FLAG 0x1
+namespace Kernel {
+static InterruptManagement* s_interrupt_management;
+static bool initialized;
+
+InterruptManagement& InterruptManagement::the()
+{
+    if (!initialized && s_interrupt_management == nullptr)
+        s_interrupt_management = new InterruptManagement(false);
+    return *s_interrupt_management;
+}
+
+AdvancedInterruptManagement& AdvancedInterruptManagement::the()
+{
+    ASSERT(s_interrupt_management != nullptr);
+    return *(AdvancedInterruptManagement*)s_interrupt_management;
+}
+
+void InterruptManagement::initialize()
+{
+    if (!initialized)
+        s_interrupt_management = new InterruptManagement(true);
+    initialized = true;
+}
+
+void InterruptManagement::switch_to_pic_mode()
+{
+    kprintf("Interrupts: Switch to Legacy PIC mode\n");
+    if (!initialized)
+        return;
+    for (auto* irq_controller : InterruptManagement::the().m_interrupt_controllers) {
+        ASSERT(irq_controller != nullptr);
+        if (irq_controller->get_model().matches("IOAPIC")) {
+            irq_controller->hard_disable();
+            dbgprintf("Interrupts: Detected %s - Disabled\n", irq_controller->get_model().characters());
+        } else {
+            dbgprintf("Interrupts: Detected %s\n", irq_controller->get_model().characters());
+        }
+    }
+}
+void InterruptManagement::switch_to_ioapic_mode()
+{
+    kprintf("Interrupts: Switch to IOAPIC mode\n");
+    if (!initialized)
```

Comment:
For these calls, should being initialized be a precondition? Seems a little strange to me for these calls to silently fail.

Reply:
I should replace it with assertion, thank you! :)